### ISSUE TITLE
Webhook: Need to extend the interface of type.Decoder

### ISSUE THREAD

**Issue by kdada** [kubernetes-sigs/controller-runtime#308]
The interface of `type.Decoder`(pkg/webhook/admission/types/types.go):
```go
// Decoder is used to decode AdmissionRequest.
type Decoder interface {
	// Decode decodes the raw byte object from the AdmissionRequest to the passed-in runtime.Object.
	Decode(Request, runtime.Object) error
}
```

The implementation is (pkg/webhook/admission/decode.go):
```go
type decoder struct {
	codecs serializer.CodecFactory
}

// NewDecoder creates a Decoder given the runtime.Scheme
func NewDecoder(scheme *runtime.Scheme) (types.Decoder, error) {
	return decoder{codecs: serializer.NewCodecFactory(scheme)}, nil
}

// Decode decodes the inlined object in the AdmissionRequest into the passed-in runtime.Object.
func (d decoder) Decode(req types.Request, into runtime.Object) error {
	deserializer := d.codecs.UniversalDeserializer()
	return runtime.DecodeInto(deserializer, req.AdmissionRequest.Object.Raw, into)
}
```

This decoder only decodes `req.AdmissionRequest.Object.Raw`. When I want to compare old object and new object in `UPDATE` request, the decoder can't decode old object for me.

So I think the decoder need an method to decode raw data:
```go
// Decoder is used to decode AdmissionRequest.
type Decoder interface {
	// Decode decodes the raw byte object from the AdmissionRequest to the passed-in runtime.Object.
	Decode(Request, runtime.Object) error
	DecodeRaw([]byte, runtime.Object) error
}
``` 

WDYT?

**Comment by FillZpp** [comment]
I have the same problem too, and i have to handle it this way for the time being:
```
fakeReq := types.Request{AdmissionRequest: &admissionv1beta1.AdmissionRequest{Object: req.AdmissionRequest.OldObject}}
p.Decode(fakeReq, oldPod)
```

**Comment by DirectXMan12** [comment]
this is addressed in #328
